### PR TITLE
[pigeon] Recursively create output target files

### PR DESCRIPTION
**Pull request by Nitin-Poojary**
Recursively creates output target files before trying to write to it.

closes #[128820](https://github.com/flutter/flutter/issues/128820)

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by tarrinneal** [comment]
@stuartmorgan do you think this is something that can be reasonably tested?

**Comment by stuartmorgan** [comment]
I don't see why not; we have lots of tests that generate output files. All we need is one that does so with a target directory that doesn't exist.

**Comment by tarrinneal** [comment]
> I'm sorry, but I didn't quite understand your point. Could you please clarify what I need to do?

Can you add a test in the pigeon_lib_test.dart file that tests writing a file in a directory that doesn't exist?

**Comment by Nitin-Poojary** [comment]
I apologize, but I am having difficulty understanding the approach required to write this test. Could you please provide guidance or instructions on how to proceed?

**Comment by Nitin-Poojary** [comment]
> I apologize, but I am having difficulty understanding the approach required to write this test. Could you please provide guidance or instructions on how to proceed?

@stuartmorgan

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
> I apologize, but I am having difficulty understanding the approach required to write this test. Could you please provide guidance or instructions on how to proceed?

To test as a unit test, you would create a temp directory (see the logic in `withTempFile` for an example) and then call `shouldGenerate` on a generator with options that would write to a non-existent subdirectory of that temp directory, then check that it was created.

Alternately it could be done as an end-to-end test [here](https://github.com/flutter/packages/blob/442d9db34717454e99cd27b188724b858d72227e/packages/pigeon/tool/shared/test_suites.dart#L405) by using an output path that doesn't exist.

File: packages/pigeon/tool/shared/test_suites.dart [hunk 1]
```diff
@@ -424,6 +424,13 @@ Future<int> _runCommandLineTests() async {
...
       tempOutput
     ],
+    // Test writing a file in a directory that doesn't exist
+    <String>[
+      '--input',
+      'pigeons/message.dart',
+      '--dart_out',
+      '$tempDir/subdirectory/message.g.dart',
+    ],
   ];
```
Comment (Nitin-Poojary) [inline packages/pigeon/tool/shared/test_suites.dart:434]: Like this?

File: packages/pigeon/tool/shared/test_suites.dart [hunk 2]
```diff
@@ -366,6 +366,13 @@ Future<int> _runCommandLineTests() async {
       '--ast_out',
       tempOutput
     ],
+    // Test writing a file in a directory that doesn't exist
```
Comment (stuartmorgan) [inline packages/pigeon/tool/shared/test_suites.dart:369]: Nit: missing final period.

File: packages/pigeon/lib/pigeon_lib.dart [hunk 3]
```diff
@@ -472,6 +473,7 @@ class DartGeneratorAdapter implements GeneratorAdapter {
     final DartOptions dartOptionsWithHeader = _dartOptionsWithCopyrightHeader(
       options.dartOptions,
       options.copyrightHeader,
+      dartOutPath: options.dartOut,
```
[the remark this answers is not in the document]
Comment (Nitin-Poojary) [reply packages/pigeon/lib/pigeon_lib.dart]: I made this change to investigate something, but then I forgot to revert it. Later, I mistakenly assumed that I had made the change intentionally, and I apologize for that.

**Review by tarrinneal — APPROVED** [review]
Thanks for putting this together!

**Comment by Nitin-Poojary** [comment]
@tarrinneal @stuartmorgan I wanted to say a big thank you for your support and guidance during the review and merging of my pull request. Your help has been invaluable, and I'm grateful for the opportunity to contribute to this project.

**Comment by tarrinneal** [comment]
> @tarrinneal @stuartmorgan I wanted to say a big thank you for your support and guidance during the review and merging of my pull request. Your help has been invaluable, and I'm grateful for the opportunity to contribute to this project.

We're always happy to help community members land good pr's! Thank you for putting in the effort and time to get this through to the finish line!

**Pull request merged**